### PR TITLE
zh-CN: translate Svelte TypeScript tutorial

### PR DESCRIPTION
**Pull request by hanyujie2002**
### Related issues and pull requests

Relates to #16965

### CODE REVIEW

**Comment by yin1999** [comment]
Please convert this PR to ready when the upstream PRs have been merged

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 1]
```diff
@@ -0,0 +1,997 @@
+---
+title: Svelte 对 TypeScript 的支持
+slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_TypeScript
+l10n:
+  sourceCommit: e9be22eaa7416206e3f263a058f0d509a7f81f88
```
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:5]:
```suggestion
  sourceCommit: 640170e669542728fc0c0b24b4c45c99ba10bfd2
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+
+很遗憾，[REPL 中目前尚不支持 TypeScript](https://github.com/sveltejs/sites/issues/156)。
+
+## TypeScript：JavaScript 的可选静态类型
+
+[TypeScript](https://www.typescriptlang.org/) 是 JavaScript 的超集，提供了可选的静态类型、类、接口和泛型等功能。TypeScript 的目标是通过其类型系统尽早发现错误，并使 JavaScript 开发更高效。其中一个重要的好处是在你输入代码时，TypeScript 使集成开发环境能够提供更丰富的环境，使发现常见错误更容易。
+
+最重要的是，JavaScript 代码也是有效的 TypeScript 代码；TypeScript 是 JavaScript 的超集。你可以直接将大多数 `.js` 文件重命名为 `.ts` 文件，它们将正常工作。
+
+我们的 TypeScript 代码能够在 JavaScript 可以运行的任何地方运行。这是如何实现的呢？TypeScript 将我们的代码“转译”为纯粹的 JavaScript。这意味着它解析 TypeScript 代码，并为浏览器生成等效的纯 JavaScript 代码以运行。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:67]:
```suggestion
我们的 TypeScript 代码能够在 JavaScript 可以运行的任何地方运行。这是如何实现的呢？TypeScript 将我们的代码“转译”为原生的 JavaScript。这意味着它解析 TypeScript 代码，并生成等效的原生 JavaScript 代码以供浏览器运行。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+
+> **备注：** 如果你对 TypeScript 如何将代码转译为 JavaScript 感兴趣，可以查看 [TypeScript 演练场](https://www.typescriptlang.org/play/?target=1&e=4#example/hello-world)。
+
+长期以来，原生 TypeScript 支持一直是 Svelte 最受欢迎的功能请求。感谢 Svelte 团队和众多贡献者的辛勤工作，我们现在有了[官方解决方案](https://svelte.dev/blog/svelte-and-typescript)，可以进行测试。在本节中，我们将向你展示如何设置带有 TypeScript 支持的 Svelte 项目，以供尝试使用。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:71]:
```suggestion
长期以来，原生 TypeScript 支持一直是 Svelte 呼声最高的特性请求。感谢 Svelte 团队和众多贡献者的辛勤工作，我们现在有了[官方解决方案](https://svelte.dev/blog/svelte-and-typescript)，可以进行测试。在本节中，我们将向你展示如何设置带有 TypeScript 支持的 Svelte 项目，以供尝试使用。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 4]
```diff
@@ -0,0 +1,997 @@
...
+
+## 为什么要使用 TypeScript？
+
+TypeScript 的主要优势包括：
+
+- 提前发现错误：编译器在编译时检查类型并提供错误报告。
+- 可读性：静态类型使代码结构性更强，使其自我说明并更易读。
+- 强大的集成开发环境支持：类型信息使代码编辑器和集成开发环境能够提供代码导航、自动补全和更智能的提示等功能。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:79]:
```suggestion
- 强大的集成开发环境支持：类型信息使代码编辑器和集成开发环境能够提供代码导航、自动补全和更智能的提示等特性。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 5]
```diff
@@ -0,0 +1,997 @@
...
+npm install
+
+npm run dev
+```
+
+## 向现有的 Svelte 项目添加 TypeScript 支持
+
+要向现有的 Svelte 项目添加 TypeScript 支持，你可以[按照这些说明](https://svelte.dev/blog/svelte-and-typescript#Adding_TypeScript_to_an_existing_project)进行操作。或者，你可以将 [`setupTypeScript.js`](https://github.com/sveltejs/template/blob/master/scripts/setupTypeScript.js) 文件下载到项目根文件夹内的 `scripts` 文件夹中，然后运行 `node scripts/setupTypeScript.js`。
+
+你甚至可以使用 `degit` 下载该脚本。这就是我们将要做的操作，开始将应用程序转换为 TypeScript。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:122]:
```suggestion
你甚至可以使用 `degit` 下载该脚本。这将是我们开始将应用程序移植到 TypeScript 所需要的操作。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 6]
```diff
@@ -0,0 +1,997 @@
...
+
+   为了修复这个问题，请将 `tabindex="-1"` 替换为 `tabindex={-1}`，像这样：
+
+   ```svelte
+   <h2 id="list-heading" bind:this="{headingEl}" tabindex={-1}>
+     {completedTodos} out of {totalTodos} items completed
+   </h2>
+   ```
+
+   这样，TypeScript 可以防止我们错误地将其分配给字符串变量。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:578]: We may remove the translation of "variable"

```suggestion
   这样，TypeScript 可以防止我们错误地将其错误地赋值为字符串。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 7]
```diff
@@ -0,0 +1,997 @@
...
+
+3. 最后，在这个文件中，我们需要为 `autofocus` 变量指定正确的类型。更新它的定义如下：
+
+   ```ts
+   export let autofocus: boolean = false;
+   ```
+
+### Todo.svelte
+
+现在，只有调用 `Todo.svelte` 组件时才会触发 `npm run check` 发出警告。让我们修复它们。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:599]:
```suggestion
现在，只有调用 `Todo.svelte` 组件时才会触发 `npm run check` 的警告。让我们修复它们。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 8]
```diff
@@ -0,0 +1,997 @@
...
+这些指令适用于任何你想要转换为 TypeScript 的 Svelte 项目。只需注意，Svelte 社区正在不断改进 Svelte 对 TypeScript 的支持，因此你应定期运行 `npm update` 来获取最新的更改。
+
+> **备注：** 如果你在 Svelte 应用程序中使用 TypeScript 时遇到任何问题，请查看[有关 TypeScript 支持的故障排除/常见问题解答部分](https://github.com/sveltejs/language-tools/blob/master/docs/preprocessors/typescript.md#troubleshooting--faq)。
+
+我们之前提到过，TypeScript 是 JavaScript 的超集，因此你的应用程序无需修改即可运行。目前，该应用程序仅是启用了 TypeScript 支持的常规 JavaScript 应用程序，而不会利用 TypeScript 提供的任何特性。你现在可以逐步添加类型。
+
+一旦你配置好了 TypeScript，你就可以通过在组件的脚本部分开头添加 `<script lang='ts'>` 来在 Svelte 组件中使用它。要通过常规的 JavaScript 文件使用它，只需将文件扩展名从 `.js` 更改为 `.ts`。你还需要更新所有相应的导入语句，将 `import` 语句中的 `.ts` 文件扩展名删除。
+
+> **备注：** 如果在 `import` 语句中使用 `.ts` 文件扩展名，TypeScript 会抛出错误，因此如果你有名为 `./foo.ts` 的文件，你必须将其导入为 `"./foo"`。
+> 更多信息请参阅 TypeScript 手册中的[捆绑程序、TypeScript 运行时和 Node.js 加载器的模块解析](https://www.typescriptlang.org/docs/handbook/2/modules.html##module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders)部分。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:152]: I found the url `https://www.typescriptlang.org/docs/handbook/2/modules.html##module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders` has been broken (the specified section is not existed), could you update this?

```suggestion
> **备注：** 如果在 `import` 语句中使用 `.ts` 文件扩展名，TypeScript 会抛出错误，因此如果你有名为 `./foo.ts` 的文件，你必须将其导入为 `"./foo"`。更多信息请参阅 TypeScript 手册中的[捆绑程序、TypeScript 运行时和 Node.js 加载器的模块解析](https://www.typescriptlang.org/docs/handbook/2/modules.html##module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders)部分。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 6 above)
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:578]:
```suggestion
   这样，TypeScript 可以防止我们将其错误地分配给字符串。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 8 above)
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:152]:
```suggestion
> 更多信息请参阅 TypeScript 手册中的[捆绑程序、TypeScript 运行时和 Node.js 加载器的模块解析](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders)部分。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 6 above)
Comment (yin1999) [reply files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md]: 这里说的是，防止将“字符串值”赋给“tabindex”这个属性（该属性接受的值是 Number 类型）。只是因为 HTML 允许使用字符串值，也允许使用数字（但实际该属性接受的字符串也应该表示一个数字）。

你建议的翻译并没有体现这个意思，请参考一下我上面给出的翻译。

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 9]
```diff
@@ -0,0 +1,997 @@
...
+这些指令适用于任何你想要转换为 TypeScript 的 Svelte 项目。只需注意，Svelte 社区正在不断改进 Svelte 对 TypeScript 的支持，因此你应定期运行 `npm update` 来获取最新的更改。
+
+> **备注：** 如果你在 Svelte 应用程序中使用 TypeScript 时遇到任何问题，请查看[有关 TypeScript 支持的故障排除/常见问题解答部分](https://github.com/sveltejs/language-tools/blob/master/docs/preprocessors/typescript.md#troubleshooting--faq)。
+
+我们之前提到过，TypeScript 是 JavaScript 的超集，因此你的应用程序无需修改即可运行。目前，该应用程序仅是启用了 TypeScript 支持的常规 JavaScript 应用程序，而不会利用 TypeScript 提供的任何特性。你现在可以逐步添加类型。
+
+一旦你配置好了 TypeScript，你就可以通过在组件的脚本部分开头添加 `<script lang='ts'>` 来在 Svelte 组件中使用它。要通过常规的 JavaScript 文件使用它，只需将文件扩展名从 `.js` 更改为 `.ts`。你还需要更新所有相应的导入语句，将 `import` 语句中的 `.ts` 文件扩展名删除。
+
+> **备注：** 如果在 `import` 语句中使用 `.ts` 文件扩展名，TypeScript 会抛出错误，因此如果你有名为 `./foo.ts` 的文件，你必须将其导入为 `"./foo"`。
+> 更多信息请参阅 TypeScript 手册中的[捆绑程序、TypeScript 运行时和 Node.js 加载器的模块解析](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders)部分。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:152]:
```suggestion
> **备注：** 如果在 `import` 语句中使用 `.ts` 文件扩展名，TypeScript 会抛出错误，因此如果你有名为 `./foo.ts` 的文件，你必须将其导入为 `"./foo"`。更多信息请参阅 TypeScript 手册中的[捆绑程序、TypeScript 运行时和 Node.js 加载器的模块解析](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-bundlers-typescript-runtimes-and-nodejs-loaders)部分。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 6 above)
Comment (hanyujie2002) [reply files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md]: "TypeScript 可以防止我们错误地将其错误地赋值为字符串" looks weird. Maybe "TypeScript 可以防止我们将其赋值为错误的字符串" is better?

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 10]
```diff
@@ -0,0 +1,996 @@
...
+
+   为了修复这个问题，请将 `tabindex="-1"` 替换为 `tabindex={-1}`，像这样：
+
+   ```svelte
+   <h2 id="list-heading" bind:this="{headingEl}" tabindex={-1}>
+     {completedTodos} out of {totalTodos} items completed
+   </h2>
+   ```
+
+   这样，TypeScript 可以防止我们将其错误地分配给字符串。
```
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:577]:
```suggestion
   这样，TypeScript 可以防止我们将其分配为错误的字符串。
```

Comment (yin1999) [reply files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:577]: 这里用“赋值”就行，分配我们一般用在“内存空间”的分配，而非“使得某个变量的值变成另一个”

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 8 above)
Comment (yin1999) [reply files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md]: ref: #32606

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md (hunk 10 above)
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:577]:
```suggestion
   这样，TypeScript 可以防止我们将其赋值为错误的字符串。
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 11]
```diff
@@ -0,0 +1,996 @@
...
+
+   为了修复这个问题，请将 `tabindex="-1"` 替换为 `tabindex={-1}`，像这样：
+
+   ```svelte
+   <h2 id="list-heading" bind:this="{headingEl}" tabindex={-1}>
+     {completedTodos} out of {totalTodos} items completed
+   </h2>
+   ```
+
+   这样，TypeScript 可以防止我们将其赋值为错误的字符串。
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:577]:
```suggestion
   这样，TypeScript 可以防止我们将其错误地赋值为字符串。
```

**Comment by hanyujie2002** [comment]
Should we convert this pr into draft and wait for upstream update? @yin1999

**Comment by yin1999** [comment]
> Should we convert this pr into draft and wait for upstream update? @yin1999

No. I believe the upstream PR will be merged shortly :)

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 12]
```diff
@@ -0,0 +1,996 @@
+---
+title: Svelte 对 TypeScript 的支持
+slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_TypeScript
+l10n:
+  sourceCommit: 640170e669542728fc0c0b24b4c45c99ba10bfd2
```
Comment (hanyujie2002) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:5]:
```suggestion
  sourceCommit: 1d5e91bbf7c4922177a9e3ae6245b0badaf1e8dd
```

File: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md [hunk 13]
```diff
@@ -0,0 +1,996 @@
+---
+title: Svelte 对 TypeScript 的支持
+slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_TypeScript
+l10n:
+  sourceCommit: 1d5e91bbf7c4922177a9e3ae6245b0badaf1e8dd
```
Comment (yin1999) [inline files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_typescript/index.md:5]:
```suggestion
  sourceCommit: 7ff3d643b325985bf4b4b756a15f2fef347e7e8c
```

**Review by yin1999 — APPROVED** [review]
Thank you @hanyujie2002

**Pull request merged**